app/src/pages/settings/CustomProviderForm.tsx

```diff
@@ -0,0 +1,873 @@
+  sdk: "AZURE_OPENAI";
+  azure_endpoint: string;
+  azure_deployment_name: string;
+  azure_api_version: string;
+  azure_auth_method: AzureAuthMethod;
+  // API Key auth
+  azure_api_key?: string;
+  // Azure AD Token Provider auth
+  azure_tenant_id?: string;
+  azure_client_id?: string;
+  azure_client_secret?: string;
+  azure_scope?: string;
+  azure_default_headers?: JSONString;
+}
+
+/**
+ * Anthropic SDK configuration.
+ */
+export interface AnthropicFormData extends BaseProviderFormData {
+  sdk: "ANTHROPIC";
+  anthropic_api_key: string;
+  anthropic_base_url?: string;
+  anthropic_default_headers?: JSONString;
+}
+
+/**
+ * AWS Bedrock SDK configuration.
+ * Uses AWS credentials for authentication.
+ */
+export interface AWSBedrockFormData extends BaseProviderFormData {
+  sdk: "AWS_BEDROCK";
+  aws_region: string;
+  aws_access_key_id: string;
+  aws_secret_access_key: string;
+  aws_session_token?: string;
+  aws_endpoint_url?: string;
+}
+
+/**
+ * Google GenAI SDK configuration.
+ */
+export interface GoogleGenAIFormData extends BaseProviderFormData {
+  sdk: "GOOGLE_GENAI";
+  google_api_key: string;
+  google_base_url?: string;
+  google_headers?: JSONString;
+}
+
+/**
+ * Discriminated union of all provider form data types.
+ * The `sdk` field serves as the discriminator for type narrowing.
+ */
+export type ProviderFormData =
+  | OpenAIFormData
+  | AzureOpenAIFormData
+  | AnthropicFormData
+  | AWSBedrockFormData
+  | GoogleGenAIFormData;
+
+export interface ProviderFormProps {
+  onSubmit: (data: ProviderFormData) => void;
+  onCancel: () => void;
+  /** Initial values for editing an existing provider. If not provided, creates a new provider with OPENAI defaults. */
+  initialValues?: ProviderFormData;
+  isSubmitting?: boolean;
+}
+
+// Form context to avoid prop drilling
+interface FormContextValue {
+  isSubmitting: boolean;
+}
+
+const FormContext = createContext<FormContextValue | null>(null);
+FormContext.displayName = "FormContext";
+
+function useFormContext(): FormContextValue {
+  const context = use(FormContext);
+  if (!context) {
+    throw new Error("useFormContext must be used within FormContext.Provider");
+  }
+  return context;
+}
+
+/**
+ * Type-safe field path helper for discriminated unions.
+ * Allows any valid field name from the union without type errors.
+ */
+type ProviderFormFieldPath = FieldPath<ProviderFormData>;
+
+// Reusable form field component - reduces Controller boilerplate
+interface FormTextFieldProps {
+  name: ProviderFormFieldPath;
+  label: string;
+  placeholder?: string;
+  description?: string;
+  isRequired?: boolean;
+  type?: "text" | "password";
+  control: Control<ProviderFormData>;
+  rules?: { required?: string };
+  flex?: boolean;
+}
+
+function FormTextField({
+  name,
+  label,
+  placeholder,
+  description,
+  isRequired,
+  type = "text",
+  control,
+  rules,
+  flex,
+}: FormTextFieldProps) {
+  const { isSubmitting } = useFormContext();
```

Comment:
this does reduce prop drilling but also it feels largely strange to mound context just for this??